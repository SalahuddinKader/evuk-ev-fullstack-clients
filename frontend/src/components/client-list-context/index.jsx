import React, { useState, useEffect, createContext } from "react";
import api from "../../api";
/**
 * Client list context display all state and functions
 * @state {array} clients
 * @state {string} search
 * @state {array} searchResult
 * @func {function} AddClientHandler
 * @func {function} retrieveClients
 * @func {function} removeClientHandler
 * @func {function} searchHandler
 */
export const ClientContext = createContext();

export const ClientListContext = (props) => {
  const [clients, setClients] = useState([]);
  const [search, setSearch] = useState("");
  const [searchResult, setSearchResult] = useState([]);

  // AddClientHandler and post clients
  const addClientHandler = async (client) => {
    const request = {
      id: Math.random() * 1000,
      ...client,
    };
    const response = await api.post("/clients", request);
    setClients([...clients, response.data]);
  };

  // Handle get api
  const retrieveClients = async () => {
    const response = await api.get("/clients");
    return response.data;
  };

  useEffect(() => {
    const getAllClients = async () => {
      const allClients = await retrieveClients();
      if (allClients) setClients(allClients);
    };
    getAllClients();
  }, []);

  // Delete individual client
  const removeClientHandler = async (id) => {
    await api.delete(`/clients/${id}`);
    setClients(clients.filter((client) => client.id !== id));
  };
  // Handling search through object value
  const searchHandler = (search) => {
    setSearch(search);
    if (search !== "") {
      const newClientList = clients.filter((client) => {
        return Object.values(client)
          .join(" ")
          .toLowerCase()
          .includes(search.toLowerCase());
      });
      setSearchResult(newClientList);
    } else {
      setSearchResult(clients);
    }
  };

  return (
    <>
      <ClientContext.Provider
        data-testid="client-list-context-provider"
        value={{
          clients,
          search,
          searchResult,
          searchHandler,
          setClients,
          addClientHandler,
          retrieveClients,
          removeClientHandler,
        }}
      >
        {props.children}
      </ClientContext.Provider>
    </>
  );
};

export default ClientListContext;
