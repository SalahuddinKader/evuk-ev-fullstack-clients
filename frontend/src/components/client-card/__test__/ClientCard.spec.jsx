import React from 'react';
import { render } from '@testing-library/react';
import { ClientContext } from '../../client-list-context';
import { BrowserRouter } from 'react-router-dom';
import ClientCard from '..';

describe('ClientCard Component', () => {
  it('matches the snapshot', () => {
    const { baseElement } = render(
      <ClientContext.Provider
        value={{
          clients: [],
          search: '',
          searchResult: [],
          searchHandler: () => {},
          setClients: () => {},
          AddClientHandler: () => {},
          retrieveClients: () => {},
          removeClientHandler: () => {},
        }}
      >
        <BrowserRouter>
          <ClientCard
            client={{
              id: 527.2950507590961,
              name: 'ClientName',
              email: 'client@example.com',
              companyName: 'ABC',
              streetAddress: '20 Courtney Road',
              postCode: 'RM164TY',
              city: 'City',
              phone: '01234567890',
            }}
          />
        </BrowserRouter>
      </ClientContext.Provider>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
