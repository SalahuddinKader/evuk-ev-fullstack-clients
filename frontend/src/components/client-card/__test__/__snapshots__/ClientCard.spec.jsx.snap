// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientCard Component matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      data-testid="client-card"
      style="max-width: 600px; margin: 0px auto 20px auto; padding: 20px 5px;"
    >
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <a
          data-testid="client-card-link"
          href="/client/527.2950507590961"
          style="text-decoration: none;"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
              data-testid="client-card-avatar"
            >
              C
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-name"
              >
                ClientName
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-email"
              >
                Email: 
                client@example.com
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-2avy51-MuiTypography-root"
                data-testid="client-card-company-fields"
              >
                Company Fields
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-company-name"
              >
                Company Name: 
                ABC
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-phone"
              >
                Phone: 
                01234567890
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-street-address"
              >
                Street: 
                20 Courtney Road
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-postcode"
              >
                Postcode: 
                RM164TY
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-city"
              >
                City: 
                City
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                data-testid="client-card-city"
              >
                Date: 
                Invalid Date
              </p>
            </div>
          </div>
        </a>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-wdtwov-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="client-card-button-delete"
            tabindex="0"
            type="button"
          >
            Delete
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
